### PR TITLE
New Cybereason plugin

### PR DESCRIPTION
**Pull request by r7-mgorny**
### Description

Describe the proposed changes:

  -

### Testing

Any testing information goes here.

## PR Requirements

Developers, verify you have completed the following items by checking them off:

### Style

Review the [style guide](https://docs.rapid7.com/insightconnect/style-guide/)

- [ ] For dependencies, pin [OS package](https://docs.rapid7.com/insightconnect/style-guide/#dockerfile) and [Python package](https://docs.rapid7.com/insightconnect/style-guide/#requirements.txt) versions
- [ ] For security, set least privileged account with ``USER nobody`` in the ``Dockerfile`` when possible
- [ ] For size, use the [slim SDK images](https://docs.rapid7.com/insightconnect/sdk-guide/#sdk-guide) when possible: ``komand/python-3-37-slim-plugin`` and ``komand/python-3-37-plugin``
- [ ] For error handling, use of [PluginException](https://docs.rapid7.com/insightconnect/error-handling-in-integrations/#plugin-exceptions) and [ConnectionTestException](https://docs.rapid7.com/insightconnect/error-handling-in-integrations#connection-exceptions)
- [ ] For logging, use [self.logger](https://docs.rapid7.com/insightconnect/sdk-guide/#logging)
- [ ] For docs, use [changelog style](https://docs.rapid7.com/insightconnect/style-guide/#changelog)
- [ ] For docs, validate markdown with ``make validate`` which calls ``mdl`` to lint ``help.md``

### Functional Checklist
- [ ] Work fully completed
- [ ] Functional
  - [ ] Any new actions/triggers include JSON [test files](https://docs.rapid7.com/insightconnect/style-guide/#tests) in the `tests/` directory created with `icon-plugin run -c sample $action > tests/$action.json`
  - [ ] Tests should all pass unless it's a negative test. Negative tests have a naming convention of `tests/$action_bad.json`
  - [ ] Unsuccessful tests should fail by raising an exception causing the plugin to die and an object should be returned on successful test
  - [ ] Add functioning test results to PR, sanitize any output if necessary
    * Single action/trigger `icon-plugin run -T tests/example.json --debug --jq`
    * All actions/triggers shortcut `icon-plugin run -T all --debug --jq` (use PR format at end)
  - [ ] Add functioning run results to PR, sanitize any output if necessary
    * Single action/trigger `icon-plugin run -R tests/example.json --debug --jq`
    * All actions/triggers shortcut `icon-plugin run -R all --debug --jq` (use PR format at end)

### Run

<details>

```
{
  "body": {
    "log": "Connect: Connecting...\nrapid7/Cybereason Plugin:1.0.0. Step name: search_for_files\n",
    "meta": {},
    "output": {
      "response": {
        "abortTime": 0,
        "abortTimeout": false,
        "actionArguments": {
          "@class": "com.cybereason.configuration.models.filesearch.FileSearchParameters",
          "fileSearchRequestConfiguration": {
            "cpuTrackingWindowMilli": 1000,
            "diskRateBytesPerMilli": 5120,
            "maxConcurrentFileSearches": 10,
            "maxDiskIOWindowMilli": 10000,
            "maxReadBytesPerFile": 110100480,
            "maxResults": 20,
            "maxYaraTimeouts": 10,
            "minFileReadPriceMilli": 1,
            "minThrottleAmountMilli": 5,
            "searchTimeoutDataScanSec": 200000,
            "searchTimeoutSec": 1200,
            "shouldUseNewAPI": false,
            "targetCpuPercentage": 20,
            "timoutPerFileScan": 9
          },
          "filters": [
            {
              "fieldName": "fileName",
              "operator": "ContainsIgnoreCase",
              "values": [
                "install.log"
              ]
            }
          ],
          "machines": []
        },
        "actionType": "FileSearchStart",
        "batchId": -671803273,
        "finalState": false,
        "globalStats": {
          "stats": {
            "AbortTimeout": 0,
            "Aborted": 0,
            "Aborting": 0,
            "AlreadyUpdated": 0,
            "BadArgument": 0,
            "ChunksRequired": 0,
            "Disconnected": 0,
            "EndedWithInvalidParam": 0,
            "EndedWithNoValidFolder": 0,
            "EndedWithSensorTimeout": 0,
            "EndedWithTooManyResults": 0,
            "EndedWithTooManySearches": 0,
            "EndedWithUnknownError": 0,
            "EndedWithUnsupportedFilter": 0,
            "EndedWithYaraCompileError": 0,
            "Failed": 0,
            "FailedSending": 0,
            "FailedSendingToServer": 0,
            "GettingChunks": 0,
            "InProgress": 0,
            "InvalidState": 0,
            "MsiFileCorrupted": 0,
            "MsiSendFail": 0,
            "NewerInstalled": 0,
            "None": 177,
            "NotSupported": 0,
            "Pending": 0,
            "Primed": 0,
            "ProbeRemoved": 0,
            "SendingMsi": 0,
            "SendingPlatform": 0,
            "Started": 0,
            "Succeeded": 0,
            "Timeout": 0,
            "TimeoutSending": 0,
            "UnauthorizedUser": 0,
            "UnknownProbe": 0,
            "partialResponse": 0
          }
        },
        "initiatorUser": "mike_rinehart@rapid7.com",
        "startTime": 1614481205231,
        "totalNumberOfProbes": 177
      }
    },
    "status": "ok"
  },
  "type": "action_event",
  "version": "v1"
}

```

<summary>
docker run --rm -i rapid7/cybereason:1.0.0 --debug run < tests.back/search_for_files.json
</summary>
</details>

### Validate

<details>

```

[*] Use ``make menu`` for available targets
[*] Including available Makefiles: ../tools/Makefiles/Helpers.mk ../tools/Makefiles/Colors.mk
--
[*] Running validators
[*] Validating plugin at .

[*] Running Integration Validators...
[*] Executing validator HelpValidator
[*] Executing validator ChangelogValidator
[*] Executing validator CloudReadyConnectionCredentialTokenValidator
[*] Executing validator RequiredKeysValidator
[*] Executing validator UseCaseValidator
[*] Executing validator SpecPropertiesValidator
[*] Executing validator SpecVersionValidator
[*] Executing validator FilesValidator
[*] Executing validator TagValidator
[*] Executing validator DescriptionValidator
[*] Executing validator TitleValidator
[*] Executing validator VendorValidator
[*] Executing validator DefaultValueValidator
[*] Executing validator IconValidator
[*] Executing validator RequiredValidator
[*] Executing validator VersionValidator
[*] Executing validator DockerfileParentValidator
[*] Executing validator LoggingValidator
[*] Executing validator ProfanityValidator
[*] Executing validator AcronymValidator
[*] Executing validator JSONValidator
[*] Executing validator OutputValidator
[*] Executing validator RegenerationValidator
[*] Executing validator HelpInputOutputValidator
[*] Executing validator SupportValidator
[*] Executing validator RuntimeValidator
[*] Executing validator VersionPinValidator
[*] Executing validator EncodingValidator
[*] Executing validator ExampleInputValidator
[*] Plugin failed validation! The following validation errors occurred:

Validator "IconValidator" failed!
        Cause: icon.png file not included in plugin.


----
[*] Total time elapsed: 1477.0049999999999ms

[*] Validating spec with js-yaml
[SUCCESS] Passes js-yaml spec check


[*] Validating markdown...
[SUCCESS] Passes markdown linting

[*] Validating python files for style...
[SUCCESS] Passes flake8 linting

[*] Validating python files for security vulnerabilities...
[main]  INFO    profile include tests: None
[main]  INFO    profile exclude tests: None
[main]  INFO    cli include tests: None
[main]  INFO    cli exclude tests: None
[main]  INFO    running on Python 3.7.3
Run started:2021-02-28 03:02:59.630318

Test results:
        No issues identified.

Code scanned:
        Total lines of code: 1824
        Total lines skipped (#nosec): 0

Run metrics:
        Total issues (by severity):
                Undefined: 0.0
                Low: 0.0
                Medium: 0.0
                High: 0.0
        Total issues (by confidence):
                Undefined: 0.0
                Low: 0.0
                Medium: 0.0
                High: 0.0
Files skipped (0):
[SUCCESS] Passes bandit security checks

[*] Checking for typos with Misspell...
[SUCCESS] Passes Misspell check

```

<summary>
make validate
</summary>
</details>

<details>

```

[*] Validating plugin with all validators at .

[*] Running Integration Validators...
[*] Executing validator HelpValidator
[*] Executing validator ChangelogValidator
[*] Executing validator CloudReadyConnectionCredentialTokenValidator
[*] Executing validator RequiredKeysValidator
[*] Executing validator UseCaseValidator
[*] Executing validator SpecPropertiesValidator
[*] Executing validator SpecVersionValidator
[*] Executing validator FilesValidator
[*] Executing validator TagValidator
[*] Executing validator DescriptionValidator
[*] Executing validator TitleValidator
[*] Executing validator VendorValidator
[*] Executing validator DefaultValueValidator
[*] Executing validator IconValidator
[*] Executing validator RequiredValidator
[*] Executing validator VersionValidator
[*] Executing validator DockerfileParentValidator
[*] Executing validator LoggingValidator
[*] Executing validator ProfanityValidator
[*] Executing validator AcronymValidator
[*] Executing validator JSONValidator
[*] Executing validator OutputValidator
[*] Executing validator RegenerationValidator
[*] Executing validator HelpInputOutputValidator
[*] Executing validator SupportValidator
[*] Executing validator RuntimeValidator
[*] Executing validator VersionPinValidator
[*] Executing validator EncodingValidator
[*] Executing validator ExampleInputValidator
[*] Executing validator ExceptionValidator
[*] Executing validator CredentialsValidator
[*] Executing validator PasswordValidator
[*] Executing validator PrintValidator
[*] Executing validator ConfidentialValidator
[*] Executing validator DockerValidator
[*] Executing validator URLValidator
[*] Plugin failed validation! The following validation errors occurred:

Validator "IconValidator" failed!
        Cause: icon.png file not included in plugin.


----
[*] Total time elapsed: 6957.992ms

```

<summary>
icon-validate --all .
</summary>
</details>

### CODE REVIEW

File: cybereason/plugin.spec.yaml [hunk 1]
```diff
@@ -0,0 +1,429 @@
...
+    required: false
+    example: '{"username": "user@example.com", "password": "mypassword"}'
+actions:
+  search_for_files:
+    title: Search for Files
+    description: Find files on any machine in your environment with a Cybereason sensor installed
+    input:
+      server_filter:
+        title: Server Filter
+        description: A Sensor filters object where you filter sensors by different criteria such as operating system
```
Comment (mrinehart-r7) [inline cybereason/plugin.spec.yaml:414]:
```suggestion
        description: A Sensor filters string for filtering sensors by different criteria such as operating system
```

File: cybereason/plugin.spec.yaml [hunk 2]
```diff
@@ -0,0 +1,429 @@
...
+        type: string
+        required: false
+        example: 'machineName: ["rapid7-windows"]'
+      file_filter:
+        title: File Filter
+        description: A fileFilters object where you filter by machine name, folder, file creation or modification time or file size with operator Equals, NotEquals, ContainsIgnoreCase, NotContainsIgnoreCase and others
```
Comment (mrinehart-r7) [inline cybereason/plugin.spec.yaml:420]:
```suggestion
        description: A fileFilters object for filtering by machine name, folder, file creation or modification time or file size with operator Equals, NotEquals, ContainsIgnoreCase, NotContainsIgnoreCase and others
```

File: cybereason/help.md [hunk 3]
```diff
@@ -0,0 +1,177 @@
...
+
+The Cybereason platform provides military-grade cyber security with real-time awareness and detection.
+
+# Key Features
+
+* Search files on machines
+
+# Requirements
+
+* Requires an username and password
```
Comment (mrinehart-r7) [inline cybereason/help.md:11]:
```suggestion
* Requires a Cybereason username and password
* Cybereason account configured as shown [here](https://nest.cybereason.com/user/login?destination=/documentation/product-documentation/191/search-and-browse-files-machines-0#pre-requisites) for File Search capabilities
```

CC @jschipp-r7 - we talked about this in our email thread with Cybereason - any thoughts?

File: cybereason/icon_cybereason/util/api.py [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+
+    def connect(self):
+        login_response = self.session.post(
+            f"{self.base_url}/login.html",
+            data={"username": self.username, "password": self.password},
+            verify=True,
+        )
+
+        if login_response.status_code not in range(200, 299):
+            raise ConnectionTestException(
```
Comment (mrinehart-r7) [inline cybereason/icon_cybereason/util/api.py:26]:
```suggestion
            raise PluginException(
```

File: cybereason/icon_cybereason/util/api.py [hunk 5]
```diff
@@ -0,0 +1,72 @@
...
+                preset=ConnectionTestException.Preset.SERVICE_UNAVAILABLE,
```
Comment (mrinehart-r7) [inline cybereason/icon_cybereason/util/api.py:27]:
```suggestion
                preset=PluginException.Preset.SERVICE_UNAVAILABLE,
```

File: cybereason/icon_cybereason/util/api.py [hunk 6]
```diff
@@ -0,0 +1,72 @@
...
+                data="There is a problem with service. Please check your credentials and permissions",
```
Comment (mrinehart-r7) [inline cybereason/icon_cybereason/util/api.py:28]:
```suggestion
                data="There is a problem connecting to Cybereason. Please check your credentials and permissions.",
```

**Review by mrinehart-r7 — CHANGES_REQUESTED** [review]

File: cybereason/icon_cybereason/util/api.py (hunk 4 above)
Comment (r7-mgorny) [reply cybereason/icon_cybereason/util/api.py:26]: This method ("connect") is used only in connection. Should we change this to PluginException and then except in connection?

**Review by mrinehart-r7 — CHANGES_REQUESTED** [review]

**Comment by mrinehart-r7** [comment]
cybereason
@r7-mgorny icon for the plugin is ready!

**Review by jschipp-r7 — DISMISSED** [review]

**Review by jschipp-r7 — APPROVED** [review]

**Pull request merged**